### PR TITLE
Parquet: Use native getRowIndexOffset support instead of calculating it

### PR DESCRIPTION
**Pull request by wypoon**

### CODE REVIEW

**Comment by wypoon** [comment]
@szehon-ho @flyrain can you please review?
cc @huaxingao

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueReader.java
Comment (Fokko) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueReader.java:32]:
```suggestion
   * @deprecated since 1.8.0, will be removed in 1.9.0 or 2.0.0; use setPageSource(PageReadStore) instead.
```

File: parquet/src/main/java/org/apache/iceberg/parquet/VectorizedReader.java
Comment (Fokko) [inline parquet/src/main/java/org/apache/iceberg/parquet/VectorizedReader.java:46]:
```suggestion
   * @deprecated since 1.8.0, will be removed in 1.9.0 or 2.0.0; use setRowGroupInfo(PageReadStore,
```

**Comment by wypoon** [comment]
@Fokko thanks for reviewing!
I'd actually updated the deprecation comments locally but failed to commit the change before I pushed.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueReader.java
Comment (wypoon) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueReader.java]: In my change, I put down 1.9.0. If there is no 1.9.0, and the methods are removed in 2.0.0 instead, I don't think that would be a problem.

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java
Comment (huaxingao) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:65]: if `pageStore.getRowIndexOffset()` is empty, does it mean [getRowIndexOffset()](https://github.com/apache/parquet-java/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ColumnChunkPageReadStore.java#L379) returns a negative value? Shall we throw Exception instead of default it to 0?

Comment (wypoon) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/ColumnarBatchReader.java:65]: That is a good question.
As I understand it, the `PageReadStore` implementation (`ColumnChunkPageReadStore`) is normally constructed with the `rowIndexOffset`, but if the offset is not available then it is constructed with -1 for the `rowIndexOffset`. `PageReadStore::getRowIndexOffset()` will not return a negative value; it will return `Optional.empty()` in that case.
I suppose we can throw an IllegalArgumentException instead in such a situation, instead of setting `rowStartPosInBatch` to 0.
@flyrain do you have an opinion on this?
Is there someone who knows Parquet well who can confirm that in normal operation, `PageReadStore::getRowIndexOffset()` should *not* return `Optional.empty()`?

File: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java
Comment (huaxingao) [inline arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java:564]: nit: Is there a better Exception than `IllegalArgumentException`? Is `IllegalStateException` a bit better?

Comment (wypoon) [reply arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java:564]: I can see why you might consider `IllegalStateException`. I do think `IllegalArgumentException` is appropriate, because the `PageReadStore` is an argument to the method being called, and the problem is with the `PageReadStore`. I think `IllegalStateException` is typically used to indicate an internal inconsistency in the module.
Consider if we use Guava's `Preconditions` to check a condition here. The condition would be that `source.getRowIndexOffset().isPresent()`. The `checkArgument` methods throw `IllegalArgumentException` and "Ensures the truth of an expression involving one or more parameters to the calling method." The `checkState` methods throw `IllegalStateException` and "Ensures the truth of an expression involving the state of the calling instance, *but not involving any parameters to the calling method*." (my emphasis)
Of course, that just expresses the opinions of the authors of Guava. There are others who might argue that `IllegalStateException` is appropriate here, or neither `IllegalStateException` nor `IllegalArgumentException`. It really doesn't matter too much. I can just throw `RuntimeException` if you do not agree with `IllegalArgumentException`.

**Review by huaxingao — APPROVED** [review]
LGTM

**Comment by wypoon** [comment]
@Fokko can you help merge this if you have no further feedback?

**Review by flyrain — APPROVED** [review]
LGTM. Thanks @wypoon for picking it up!

**Comment by huaxingao** [comment]
@Fokko @flyrain Should we merge this PR? I am waiting for it to be merged so I can clean up my [temporary code](https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/BaseBatchReader.java#L90)

**Comment by flyrain** [comment]
I will merge it if there is no new comment by EOD.

**Review by szehon-ho — COMMENTED** [review]
Very small comment, lgtm otherwise

**Review by szehon-ho — APPROVED** [review]

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueReader.java
Comment (wypoon) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueReader.java]: Done.

**Comment by flyrain** [comment]
Thanks @wypoon for working on it. Thanks @huaxingao @Fokko @szehon-ho for the review.

**Pull request merged**
